### PR TITLE
Decompose addmm with Gemm | feat(torchlib)

### PR DESCRIPTION
**Pull request by justinchuby**
Decompose addmm with Gemm by creating a special variant for `FLOAT` and conditionally check for the ranks if the input tensors. The if branch is expected to be folded away by constant folding passes.

I have not found other instances where Gemm is used in the torch.onnx exporter.

Fixes https://github.com/microsoft/onnxscript/issues/1089

cc @baijumeswani

### CODE REVIEW

File: onnxscript/function_libs/torch_lib/graph_building.py [hunk 1]
```diff
@@ -282,6 +282,7 @@ def eval_function(  # type: ignore[override]
             param = name_to_schema[name]
             # Cast int to float if needed
             if param.type in {float, "float"}:
+                print(name, param.type)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/graph_building.py:285]:
```suggestion
```

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 2]
```diff
@@ -222,14 +235,42 @@ def aten_addcmul(
...
+    alpha: float = 1.0,
+) -> TRealUnlessFloat32:
     """addmm(Tensor self, Tensor mat1, Tensor mat2, *, Scalar beta=1, Scalar alpha=1) -> Tensor"""
 
     mat1_mat2 = op.MatMul(mat1, mat2)
     scaled_mat1_mat2 = op.Mul(mat1_mat2, alpha)
     scaled_self = op.Mul(self, beta)
-    return op.Add(scaled_self, scaled_mat1_mat2)
+    result = op.Add(scaled_self, scaled_mat1_mat2)
+    return result
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:250]:
```suggestion
    return op.Add(scaled_self, scaled_mat1_mat2)
```

**Comment by BowenBao** [comment]
`aten::mm` is also exported as Gemm by torchscript exporter. Although I'm not sure if it is better to export as Gemm or Matmul in this case.

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 3]
```diff
@@ -232,6 +249,29 @@ def aten_addmm(
     return op.Add(scaled_self, scaled_mat1_mat2)
 
 
+@torch_op("aten::addmm")
+def aten_addmm_gemm(
+    self: FLOAT, mat1: FLOAT, mat2: FLOAT, beta: float = 1.0, alpha: float = 1.0
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_lib/ops/core.py:254]: Why only do 'gemm' variant on float inputs?

**Comment by justinchuby** [comment]
> `aten::mm` is also exported as Gemm by torchscript exporter. Although I'm not sure if it is better to export as Gemm or Matmul in this case.

Yeah Gemm seems to be more specialized for that case. Backends should just do the right thing for matmul imo

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 3 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:254]: ORT implements Gemm for float inputs only. So I set it accordingly for practicality

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py]: Changed to TFloat because it supports float16 as well

**Review by BowenBao — APPROVED** [review]
LGTM, thanks!

**Pull request merged**